appsrc/dcfile/dcbriggs.man
.TH DCBRIGGS 1 "15 February 2000" "DICOM PS3" "Create DICOM Briggs pattern"
.SH NAME
dcbriggs \- ACR/NEMA DICOM PS3 ... Create DICOM Briggs pattern
.SH SYNOPSIS
.HP 10
.B dcbriggs
[
.B \-columns|\-cols|\-x " columns"
]
[
.B \-rows|\-y " rows"
]
[
.B \-bits|\-depth|\-d " bits"
]
[
.B \-minval " i"
]
[
.B \-maxval " i"
]
[
.B \-signed
]
[
.B \-inverted|\-invert|\-monochrome1
]
[
.B \-contrast " value"
]
[
.B \-surround " percent"
]
[
.B \-background " percent"
]
.so man1/optout.so
.SH DESCRIPTION
.LP
.B dcbriggs
writes a grayscale MONOCHROME2 or MONOCHROME1 Briggs 8-target pattern (BPT#4)
as a DICOM
Secondary Capture Image Storage SOP Class instance stored in a DICOM file. The
format, rationale, use and scoring of a Briggs pattern are described in Boeing Technical
Report D180-25066-1 by S.J. Briggs, January 1979. Basically a set of checkerboard
patterns of varying spatial and contrast resolution are created in order to visually
evaluate the quality of the system displaying the pattern.
.LP
During the creation, specific attributes may be deleted, added or
replaced, class and instance unique identifiers generated or removed,
group lengths added, or private attributes removed, as specified by
the replacement options described in dcintro(1). This allows the
specification of additional DICOM attributes, such as the patient's
name, necessary to make a complete and useful SC instance.
.SH OPTIONS
The verbose output goes to standard error.
.PP
.PP
The basic switches are described in dcintro(1). Options specific to this program are:
.TP
.BI \-columns|\-cols|\-x " columns"
.RS
Defaults to 512.
.RE
.TP
.BI \-rows|\-y " rows"
.RS
Defaults to 512.
.RE
.TP
.BI \-bits|\-depth|\-d " bits"
.RS
Defaults to 8.
.RE
.TP
.BI \-minval " i"
.RS
Defaults to 0. Specifies the pixel value to store for the 0% gray value in the Briggs pattern.
.RE
.TP
.BI \-maxval " i"
.RS
Defaults to (2^bits)-1. Specifies the pixel value to store for the 100% gray value in the Briggs pattern.
(Note this does not have to be (2^bits)-1).
.RE
.TP
.BI \-signed
.RS
Treat the minval and maxval as signed and set the Pixel Representation
in the DICOM image to signed.
.RE
.TP
.BI \-inverted|\-invert|\-monochrome1
.RS
Invert the grayscale range and set the Photometric Interpretation
to MONOCHROME1 (zero is white) rather than the default MONOCHROME2 (zero
is black). The image should display the same in a DICOM viewer that
correctly handles Photometric Interpretation.
.RE
.TP
.BI \-contrast " value"
.RS
Optional. Specifies the target command contrast between light and dark checkers. Defaults
to a value of 7. This corresponds to Briggs' C value.
.RE
.TP
.BI \-surround " percent"
.RS
Optional. Specifies percent of maxval-minval used for the target surround. Should be an
integer. Divided by 100, this corresponds to Briggs' S value. Defaults to 70%.
.RE
.TP
.BI \-background " percent"
.RS
Optional. Specifies percent of maxval-minval used for the background and spacing
between targets. Should be an
integer. Defaults to 30%.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
\ 
.RE
# Create an 8 bit pattern from 0 to 255, ie. BPT#4(7,.7,256)
.RE
# and make a PGM and a GIF file from it using netpbm tools  ...
.RE
% dcbriggs -bits 8 -minval 0 -maxval 255 briggs.dc3
.RE
% dctopnm  briggs.dc3 briggs.pgm
.RE
% ppmtogif briggs.pgm > briggs.gif
.RE
\ 
.RE
# Create a 12 bit pattern from 0 to 4095 with a C of 112 and S of .55, ie. BPT#4(13,.55,4096) ...
.RE
% dcbriggs -bits 12 -minval 0 -maxval 4095 -contrast 112
.RE
\ 
.RE
# Create a 16 bit pattern from -1024 to 3071 with a C of 112 and S of .55 ...
.RE
% dcbriggs -bits 16 -minval ' -1024' -maxval 3071 -signed -contrast 112 -surround 55
.RE
\ 
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dccp(1) ,
.BR dciodvfy(1) ,
.BR dcsmpte(1) ,
.BR dcintro(1)
.SH AUTHOR
Copyright (C) 1993-2010. David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
111219:	libsrc/src/dctool/attrmxrd.cc: check for HighBit being more than one less than BitsStored and correct BitsStored accordingly (otherwise will create zero pixel values on copy) (000408)

111213:	appsrc/dcfile/dcuncat.cc: add option to replace instance number starting with a specific number to allow scripts to extend a series with successive invocations

111212:	libsrc/standard/condn.tpl/module,strval/base.tpl,dx.tpl: improve animal versus human conditions to check SRT homo sapiens, and make view position and body part examined human versus animal specific

111212:	libsrc/standard/elmdict/other.tpl: add more Sound private elements

111117:	libsrc/standard/strval/base.tpl: add CSD for 99NCIAIM

111109:	appsrc/dcfile/dcuncat.cc: add options to replace instance number, frame increment pointer and frame increment pointer vectors, as well as add number of frames during unenhance based on SOP class

111106:	libsrc/standard/elmdict/other.tpl: add Vepro private elements

111031:	libsrc/standard/elmdict/other.tpl: add CTP private elements

111030:	libsrc/standard/elmdict/other.tpl: add Segami private elements

111029:	libsrc/standard/elmdict/dicom3.tpl: add (restore from trial) Observation UID (per CP 1147)

111028:	libsrc/standard/elmdict/other.tpl: add NeuroQuant private elements

111024:	libsrc/standard/elmdict/other.tpl: update fuji private attributes with descriptions

110918:	libsrc/standard/elmdict/diconde.tpl: update with new or modified data elements with cleanup from Patrick

110918:	libsrc/standard/elmdict/siemens.tpl: update some private elements based on Artis conformance statement

110916:	libsrc/standard/elmdict/gems.tpl: resolve inconsistencies in GE PET conformance statement VRs SS/SL/FL for private elements by choosing those most commonly seen in the field (still some inconsistencies between models)

110904:	libsrc/standard/elmdict/camtron.tpl: add private data elements encountered in DICOMDIR, as well as QCA private data elements from GEMS/Camtronics CRS conformance statement

110901:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Sup 152 FT ophthalmic thickness map (dictionary and SOP class only; not IOD yet)

110822:	libsrc/standard/elmdict/dicom3.tpl: update data dictionary (only) for CPs for FT Aug 2011 WG 6 meeting

110714:	libsrc/standard/elmdict/dicom3.tpl: remove 'unified' from names of various attributes (per Sup 96 FT)

110716:	libsrc/standard/strval/base.tpl: add 99PMP CSD

110714:	libsrc/standard/elmdict/dicom3.tpl: exposure index etc are DS not FD (CP 1024)

110709:	libsrc/standard/elmdict/dicom3.tpl: change (0018,9430) from US to FL per Sup 139

110612:	libsrc/standard/elmdict/diconde.tpl: update for DICONDE eddy current standard

110612:	libsrc/standard/elmdict/dicondep.tpl: correct case of DICONDE DX (Dx) private owner for E2699

110609:	libsrc/standard/elmdict/dicom3.tpl: add Sup 151 FT data elements

110608:	libsrc/standard/elmdict/elscint.tpl: update Elscint private elements

110531:	libsrc/standard/elmdict/other.tpl: update Hitachi CT private elements

110531:	libsrc/standard/elmdict/siemens.tpl: update mammo private elements

110529:	libsrc/standard/condn.tpl,iodcomp/dx.tpl,module/base.tpl,dx.tpl,xaxrf.tpl,strval/dx.tpl,xaxrf.tpl: add breast tomosynthesis IOD, refactoring 3D XRay accordingly

110529:	libsrc/standard/elmdict/other.tpl: more private data elements for Hologic 

110528:	appsrc/dcfile/dciodvfy.cc: do not complain about SCOORD without child IMAGE when SELECTED FROM is by reference (000406)

110527:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: check that references to instances in content tree are in top level evidence sequences in SR (000405)

110526:	libsrc/standard/elmdict/gems.tpl: add names for CT iterative reconstruction private data elements

110522:	libsrc/src/dctool/attrothr.cc: do not throw assertion if computed length is equal (which may occur for single bit images)

110522:	libsrc/standard/iodcomp/base.tpl,libsrc/standard/condn.tpl: permit FoR Module in Segementation (and do CP to standard to allow may be present otherwise)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect SCOORD without child IMAGE (000403)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect quotes in code meaning of measurement units (000404)

110522:	libsrc/src/dctool/attrtypv.cc,locale/mesgtext.tpl: permit PN with multiple carets in multiple component groups (000402)

110521:	libsrc/standard/elmdict/acuson.tpl,agfa.tpl,isg.tpl,other.tpl,siemens.tpl,strval/base.tpl: more private elements and coding scheme designators

110517:	libsrc/standard/module/base.tpl: correct type (1->2) and number of items (1->0-n) in SourceImageSequence in DerivationImageMacro

110517:	libsrc/standard/elmdict/siemens.tpl: yet more syngo private elements

110516:	libsrc/standard/cond.tpl: no longer require version for UCUM codes (CP 1060)

110516:	libsrc/standard/strval/base.tpl: add 99_MI private coding scheme designator

110516:	libsrc/include/dctool/attr.h,attrseq.h,generic/listsimp.h,src/dctool/attrseq.cc,attrvrfy.cc,standard/condn.tpl,module/softcopy.tpl,sr.tpl,support/module.awk:
		though VM for sequences number of items was being specified in modules, it was not being applied;
		apply it (000256), as well as make it work (since number of items was not being computed correctly
		in SequenceAttribute::verifyVM()); also add support for conditional check of VM, for example for
		GraphicData depending on Grahic Type (000401)

110516:	libsrc/standard/elmdict/toshiba.tpl: update CT private elements from conformance statement for V4.70 or later version

110514:	libsrc/standard/elmdict/siemens.tpl: add more syngo private elements, and update guessed VRs of existing ones with observed explicit VRs

110511:	libsrc/standard/elmdict/gems.tpl,toshiba.tpl: add more encountered explicit VR private elements

110511:	libsrc/standard/elmdict/other.tpl: update/add McKesson private elements from conformance statement


appsrc/dcfile/dcbriggs.cc
#include "attrtype.h"
#include "attrothr.h"
#include "attrmxls.h"
#include "transynu.h"
#include "elmconst.h"
#include "mesgtext.h"
//#include "ioopt.h"
#include "dcopt.h"

#include "briggsrc.h"

int
main(int argc,char **argv)
{
	bool bad=false;

	GetNamedOptions 	options(argc,argv);
	DicomOutputOptions 	dicom_output_options(options);

	bool verbose=options.get("verbose") || options.get("v");

	unsigned rows=512;
	(void)(options.get("rows",rows) || options.get("x",rows));

	unsigned cols=512;
	(void)(options.get("cols",cols) || options.get("columns",cols) || options.get("y",cols));

	unsigned bits=8;
	(void)(options.get("bits",bits) || options.get("depth",bits) || options.get("d",bits));

	if (bits < 1 || bits > 16) {
		cerr << EMsgDC(OptionUnsupported) << " - bits = " << dec << bits << endl;
		bad=true;
	}

	int minval=0;
	(void)options.get("minval",minval);

	int maxval=(int)(((long)(0x0001l)<<bits)-1);
	(void)options.get("maxval",maxval);

	bool wantsigned=options.get("signed");

	bool wantinverted=options.get("invert") || options.get("inverted") || options.get("monochrome1");

	unsigned contrast=7;
	(void)options.get("contrast",contrast);
	Assert(maxval-minval >= contrast);

	unsigned surround=70;
	(void)options.get("surround",surround);
	Assert(surround <= 100);

	unsigned background=30;
	(void)options.get("background",background);
	Assert(background <= 100);

	dicom_output_options.done();
	options.done();

	DicomOutputOpenerFromOptions output_opener(
		options,dicom_output_options.filename,cout);

	cerr << dicom_output_options.errors();
	cerr << options.errors();
	cerr << output_opener.errors();

	if (!dicom_output_options.good()
	 || !options.good()
	 || !output_opener.good()
	 || !options
	 || bad) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_output_options.usage()
			<< " [-rows|-y rows]"
			<< " [-cols|-columns|-x columns]"
			<< " [-bits|-depth|-d bits]"
			<< " [-minval i]"
			<< " [-mavxal i]"
			<< " [-signed]"
			<< " [-inverted|-invert|-monochrome1]"
			<< " [-contrast value]"
			<< " [-surround percent]"
			<< " [-background percent]"
			<< " [-v|-verbose]"
			<< " [" << MMsgDC(OutputFile) << "]"
			<< " >" << MMsgDC(OutputFile)
			<< endl;
		exit(1);
	}

	DicomOutputStream dout(*(ostream *)output_opener,
		dicom_output_options.transfersyntaxuid,
		dicom_output_options.usemetaheader,
		dicom_output_options.useimplicitmetaheader,
		dicom_output_options.addtiff);

	ManagedAttributeList list;

	TransferSyntax transfersyntax(
		dicom_output_options.transfersyntaxuid
		? dicom_output_options.transfersyntaxuid
		: DefaultTransferSyntaxUID);

	TextOutputStream log(cerr);

	Uint16 bitsAllocated=((bits-1)/8+1)*8;
	Uint16 bitsStored=bits;
	Uint16 highBit=bits-1;

	list+=new UnsignedShortAttribute(TagFromName(BitsAllocated),bitsAllocated);
	list+=new UnsignedShortAttribute(TagFromName(BitsStored),bitsStored);
	list+=new UnsignedShortAttribute(TagFromName(HighBit),highBit);
	list+=new UnsignedShortAttribute(TagFromName(Rows),rows);
	list+=new UnsignedShortAttribute(TagFromName(Columns),cols);
	list+=new UnsignedShortAttribute(TagFromName(SamplesPerPixel),1u);
	list+=new UnsignedShortAttribute(TagFromName(PixelRepresentation),wantsigned?1u:0u);
	list+=new CodeStringAttribute(TagFromName(PhotometricInterpretation),wantinverted ? "MONOCHROME1":"MONOCHROME2");

	// various Type 1 and Type 2 attributes for mandatory SC modules ...

	list+=new PersonNameAttribute(TagFromName(PatientName),"^^^^");
	list+=new LongStringAttribute(TagFromName(PatientID));
	list+=new DateStringAttribute(TagFromName(PatientBirthDate));
	list+=new CodeStringAttribute(TagFromName(PatientSex));
	list+=new ShortStringAttribute(TagFromName(StudyID));
	list+=new DateStringAttribute(TagFromName(StudyDate));
	list+=new TimeStringAttribute(TagFromName(StudyTime));
	list+=new PersonNameAttribute(TagFromName(ReferringPhysicianName),"^^^^");
	list+=new ShortStringAttribute(TagFromName(AccessionNumber));
	list+=new IntegerStringAttribute(TagFromName(SeriesNumber));
	list+=new IntegerStringAttribute(TagFromName(InstanceNumber));

	list+=new CodeStringAttribute(TagFromName(Modality),"OT");
	list+=new CodeStringAttribute(TagFromName(ConversionType),"WSD");

	list+=new LongStringAttribute(TagFromName(Manufacturer));
	list+=new CodeStringAttribute(TagFromName(PatientOrientation));

	Briggs_PixelDataSource *pixeldatasrc=new Briggs_PixelDataSource(rows,cols,wantsigned,wantinverted,(Uint16)minval,(Uint16)maxval,
						contrast,surround,background);
	Assert(pixeldatasrc);

	list+=new OtherUnspecifiedLargeAttribute(
		TagFromName(PixelData),
		pixeldatasrc,
		rows,
		cols,
		1,
		1,
		&transfersyntax,
		0,bitsAllocated,bitsStored,highBit);

	if (!usualManagedAttributeListWrite(list,dout,
		dicom_output_options,log,verbose)) return 1;

	return 0;
}

